rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,221 @@
+
+We will continue to use the new development and deployment pipeline created for the trials. Much of the functionality is defined in [a new `Deploy_App_Downstream` Jenkins job](https://github.com/alphagov/govuk-puppet/blob/c24ff191ce7fc8a38ceb464f0139faba04b9734b/modules/govuk_jenkins/templates/jobs/deploy_app_downstream.yaml.erb), which works like this:
+
+1. **Integration**
+
+   1. Deployment runs.
+   1. [**new**] Wait for the Smoke tests for the app to pass [[1](https://github.com/alphagov/govuk-puppet/pull/10387)].
+   1. [**new**] Remotely trigger the same deployment in Staging.
+
+1. **Staging**
+
+   1. [**new**] Repeat as for Integration, but to Production.
+
+We propose to apply this automation across our apps. Given the scale of the change, we need to think carefully about how we can make it consistently safe and secure, before we enable CD for an app.
+
+### Safety concerns
+
+The person who merges a PR must be confident that it is safe to do so. As part of the trial of CD, [we added PR templates to each affected app](https://github.com/alphagov/smart-answers/pull/4758), to explicitly warn developers about the automatic deployment.
+
+However, we also want to ensure that a faulty PR merged in haste can be caught before it impacts the live site. There are a number of checks we can perform to help ensure this, without aiming for perfection.
+
+> **We won't check alerts.** At the time of writing, we have no evidence that checking Icinga alerts has any advantage over the checks discussed below. We can investigate such a check if we find evidence [[1](https://github.com/alphagov/govuk-puppet/pull/10597)].
+
+> **We will only consider _live_ code.** Interactions due to manual and scheduled tasks can be checked manually, using a branch deployment. In the worst case, we should have ample time to fix any faults that develop.
+
+#### Check: Internal repo tests pass (i.e. features are tested in detail)
+
+For the purpose of automating deployments:
+
+- We will use code coverage as our indicator of previous, thorough testing. While an app may have a [rigorous testing strategy](https://github.com/alphagov/content-publisher/blob/69bed3535de1e65bda9d5909a3bc56a319e46576/docs/testing-strategy.md), it's hard to know if this has been applied consistently as the app has evolved.
+
+- We will require a minimum level of coverage that is realistic for all of our apps. We will not pursue this once CD is enabled, as doing so could encourage [a lack of rigour in the way we write our tests](https://jeroenmols.com/blog/2017/11/28/coveragproblem/).
+
+- We will not check coverage of periphery code e.g. CSS, JavaScript. This is because it has a low footprint in our codebases, and it's hard to measure coverage for: the frameworks we use don't support it.
+
+An app has "enough" of these tests when:
+
+- Its code coverage exceeds 95% at the point when CD is enabled.
+
```

Comment:
Is there some rationale for the 95 number? A quick (potentially flawed) search of your file seemed to indicate we have ~50% of apps that don't meet this number. Which would indicate that we've got a bunch of fixing to do to meet this standard, whereas ~80% meet a 90% standard. Given coverage is quite a blunt number I was wondering whether we had evidence to suggest the apps under the threshold were under-tested.

It does seem prudent to have something that indicates a level of quality even it is arbitrary and it doesn't seem an easy task so big kudos for the thinking and research behind it.

Anecdotally I was surprised how bad Content Store fared (68%) given its simplicity. Looking at the coverage it seems to be all to non-application code, which - although good to clean up - seems to have limited impact on the difference between it being automatically and manually deployed. So I wondered if the 95% might sign people up to a bunch of clean-up that doesn't affect much app code.

Reply:
The number is indeed, arbitrary. I chose 95% because most of our repos are very near to meeting it, but not quite. I think this is important: we should have something to motivate us to look at the tests we have already, and think about whether any are missing. In my exploration of this among our apps, I've found:

- Some with unused code. This is low effort to delete, so not a good reason to lower the threshold.
- Lots with untested rake tasks. These can be ignored, which is a caveat I make [later on](https://github.com/alphagov/govuk-rfcs/blob/continuous-deployment/rfc-128-continuous-deployment.md#automatic-deployments).

I wouldn't expect work to improve coverage to take more than a day, for most apps. Apart from requiring less effort, do you see any advantage in lowering the threshold to accommodate more apps?

Reply:
Sure, I get you. So there is an aspect of motivation in a number that is slightly out of reach to encourage a review of most test suites before enabling CD. That seems reasonable.

As discussed in the meeting today it'd be good to capture the spirit of what the 95% means, which I guess is thorough unit tests and integration tests of functionality that exercise code within and outside of the coverage (HTML, JS)

Reply:
Sure, thanks for reminding me. I've revised the first bullet point around the reasoning for this check in https://github.com/alphagov/govuk-rfcs/pull/128/commits/ee95fef20edb6d395a19545daeeab3c0508adfa4.